### ISSUE TITLE
[2.35] - mockClient running middlewares which fails matching to the mock store

### ISSUE THREAD

**Issue by klippx** [tulios/mappersmith#247]
In PR #168 we added headers to mock strategy. Now I have an issue in 2.35 release that my mocks are no longer matching because of `DurationMiddleware` running and setting `x-started-at` to a value that is later unable to be matched in `isExactMatch` (specifically in the `headerMatch`):
```
    {
      A: 'correlation-id=f497e1b8-0fc8-42c5-87e3-5ef58f39bd8e&service-to-service-request=yes&x-started-at=1637663489675',
      B: 'correlation-id=f497e1b8-0fc8-42c5-87e3-5ef58f39bd8e&service-to-service-request=yes&x-started-at=1637663489668'
    }
```

A workaround is to do a `MockDate.set(date)` in a beforeEach - which felt a bit random - but without it the Date.now() values will be mismatching and failing to match the correct mock. Mocks not matching have always been a bit hard to debug, and now it is even harder.

This was unexpected though: We do not provide headers, so the test should just work, right? 
- Yes, but however, the problem is the fact that `mockClient` (as per specification!) runs the middleware stack on the `mock` gateway! This is different to how `mockRequest` works. So actually, all tests work if switched to `mockRequest` API instead (without the need for calling `MockDate.set(date)`).

In summary:
- mockClient runs middlewares which potentially adds headers
- This can cause existing tests to fail (if they are using mockClient) even if they are not providing headers to the mock definition
- #168 was meant to be backwards compatible _if_ your mock definition did not add headers and break if there was mismatch only
- #168 got stabbed in the back by `mockClient` running middlewares which we were unaware of

Any opinions what we should do, a few options:
- We can proceed, add warnings about the differences mockClient and mockRequest to the README
- We can add best practices about using MockDate.set when running tests (and using DurationMiddleware)
- We can try to make this new behaviour more optional
- We can revert the whole thing

WDYT? Discuss.

**Comment by tulios** [comment]
I think we should improve the ergonomics of the `DurationMiddleware`, maybe use a fixed test date when in test mode. Or ignore the headers in test mode. Users should test user land code and not mappersmith, so I don't see a lot of value in testing auto included headers, wdyt @klippx ?

**Comment by klippx** [comment]
> ignore the headers in test mode.

I like it, for example:

```diff
  isExactMatch (request) {
+    // (omitted)

+    const relevantHeaders = { ...this.headersObject }
+    delete relevantHeaders['x-started-at']

    const headerMatch = !this.headers || (
      this.headersFunction
        ? this.headers(request.headers())
-        : isSubset(this.headersObject, request.headers())
+        : isSubset(relevantHeaders, request.headers())
    )
  },
```

Something like this would completely solve my issue. If we do this then we can forget all about `MockDate.set`, readme updates and whatnot. My somewhat painful upgrade to 2.35 then suddenly becomes a one liner `.set('correlation-id', '123')` in one single spec out of 768 in total 👍

**Comment by Yurickh** [comment]
Is there any way to track which headers have been included by client middleware, and which have been manually provided in the mock definitions?
I think it'd make sense to take only the latter into consideration when matching mocks, to ensure the mock definition in the test body is a good reflection of what is going to be used as a matcher.

On the other hand, I do believe this problem also applies to parameters and body. A middleware that adds a parameter, for example, might make matching fail even though it's nowhere defined in the mock definition inside the test.

I'd love to understand _why_ we run the client middleware on top of the mock definition 🤔 it came as a surprise for me in the first place to be honest.

**Comment by klippx** [comment]
There was also this approach suggested which is a lot cleaner and also solves the problem:

```diff
@ src/middleware/duration.js:4 @
/**
 * Adds started_at, ended_at and duration headers to the response
 */
- const DurationMiddleware = () => ({
+ const DurationMiddleware = ({ mockRequest }) => ({
  prepareRequest (next) {
+    if (mockRequest) {
+      return next()
+    }
    return next().then(request => request.enhance({
      headers: { 'X-Started-At': Date.now() }
    }))
  }
```

With this approach `DurationMiddleware` take responsibility for its own ergonomics.

**Comment by Yurickh** [comment]
I don't think this problem is limited to DurationMiddleware. From my perspective, letting middleware change what is going to be matched is very unintuitive and it's what ultimately leads to this issue.

Say I have a middleware that injects a param to outgoing requests called `count` in my client that increments whenever the middleware is run.

If my mock definition is
```ts
mockClient(client)
  .resource('Potato')
  .method('solid')
  .with({
    someParameter: 'is here'
  })
```

I don't see why we'd take `count` into consideration when trying to match outgoing requests. They should always be compared using what's in your actual mock definition, in this example, the `someParameter` param.

On this example as well, we might even be unable to match a request, since the `count` parameter will be ever increasing.

**Comment by klippx** [comment]
I appreciate the concern, but this issue is specifically about what I observed in consumer land when upgrading to 2.35 and as far I am concerned this issue can be solved by fixing DurationMiddleware (see #248)

I think that I agree with what you said, but I haven't seen that this is an actual production issue yet. If it is, the right move would be to open a new issue for that and provide a breaking example that proves the point and we can discuss potential fixes there.

**Comment by tulios** [comment]
@Yurickh a good idea is to add a canonical way for middleware to add headers, so we can track and differentiate between input headers from userland. I think it's a good idea.

**Comment by Yurickh** [comment]
@tulios Would this also apply to anything that middleware can modify? Like body and params? This might be a more apparent issue because middleware are more likely to inject headers than anything else, but I do think this issue affects all of the api surface of a request.

**Comment by tulios** [comment]
I think we can diff the input/output. We could do it for all properties. A bit harder with things like body, but we can still know if a middleware changed it. This change has the potential to improve error messages, etc.
